### PR TITLE
Rule S3603: Methods with "Pure" attribute should return a value

### PR DESCRIPTION
**Pull request by valhristov**

### CODE REVIEW

File: sonaranalyzer-dotnet/rspec/vbnet/Sonar_way_profile_no_hotspot.json [hunk 1]
```diff
@@ -5,8 +5,12 @@
                      "S112",
                      "S114",
                      "S117",
+                     "S1048",
```
Comment (valhristov) [inline sonaranalyzer-dotnet/rspec/vbnet/Sonar_way_profile_no_hotspot.json]: This is generated by rspec.ps1, I guess this PR was the first to run it with the new changes we made yesterday

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/RspecStrings.resx [hunk 2]
```diff
@@ -3319,7 +3319,7 @@
     <value>Minor Security Hotspot</value>
   </data>
   <data name="S2255_Description" xml:space="preserve">
-    <value>Using cookies is security-sensitive. It has led in the past to the following vulnerabilities:</value>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/RspecStrings.resx]: Please undo these changes (or move them to a separate PR).

File: sonaranalyzer-dotnet/rspec/vbnet/Sonar_way_profile_no_hotspot.json (hunk 1 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/rspec/vbnet/Sonar_way_profile_no_hotspot.json]: As we have other changes which don't belong to this PR, I think it's better to have only the rule S3603 added here.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/PureAttributeOnVoidMethod.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class PureAttributeOnVoidMethod : PureAttributeOnVoidMethodBase
+    {
+        protected override DiagnosticDescriptor Rule { get; } =
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/PureAttributeOnVoidMethod.cs]: Not sure of the real impact but for all rules we have a static field so that this creation is not done every time. If we think this is equivalent then we should do a PR to update all classes.

Bottom line, please use the static field here.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs]: AFAIK we always have put the rules in `SonarAnalyzer.Rules.Common`.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PureAttributeOnVoidMethod.cs [hunk 5]
```diff
@@ -31,39 +27,9 @@ namespace SonarAnalyzer.Rules.CSharp
...
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     [Rule(DiagnosticId)]
-    public sealed class PureAttributeOnVoidMethod : SonarDiagnosticAnalyzer
+    public sealed class PureAttributeOnVoidMethod : PureAttributeOnVoidMethodBase
     {
-        internal const string DiagnosticId = "S3603";
-        private const string MessageFormat = "Remove the 'Pure' attribute or change the method to return a value.";
-
-        private static readonly DiagnosticDescriptor rule =
+        protected override DiagnosticDescriptor Rule { get; } =
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PureAttributeOnVoidMethod.cs]: See comment on the VB.Net class.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs (hunk 4 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/PureAttributeOnVoidMethodBase.cs]: Ok all namespaces in the common don't make sense... Let's fix that later.
